core/src/main/java/io/grpc/LoadBalancer2.java

```diff
@@ -181,12 +206,59 @@ private PickResult(Subchannel subchannel, Status status) {
-     * Subchannel will not fail the RPC, but will only leave it buffered.
+     * A decision to proceed the RPC on a Subchannel.
      *
      * <p>Only Subchannels returned by {@link Helper#createSubchannel} will work.  DO NOT try to
      * use your own implementations of Subchannels, as they won't work.
+     *
+     * <p>When the RPC tries to use the return Subchannel, which is briefly after this method
+     * returns, the state of the Subchannel will decide where the RPC would go:
+     *
+     * <ul>
+     *   <li>READY: the RPC will proceed on this Subchannel.</li>
+     *   <li>IDLE: the RPC will be buffered.  Subchannel will attempt to create connection.</li>
+     *   <li>All other states: the RPC will be buffered.</li>
+     * </ul>
+     *
+     * <p>Note that Subchannel's state may change at the same time the picker is making the
+     * decision, which means the decision may be made with (to-be) outdated information.  For
+     * example, a picker may return a Subchannel known to be READY, but it has become SHUTDOWN when
+     * used by RPC, which makes the RPC to be buffered.  The LoadBalancer will soon learn about
+     * the Subchannels' transition from READY to SHUTDOWN, create a new picker and allow the RPC to
+     * use another READY transport if there is any.
+     *
+     * <p>You will want to avoid running into a situation where there are READY Subchannels out
+     * there but some RPCs are still buffered for longer than a brief time.
+     * <ul>
+     *   <li>This can happen if you return Subchannels with states other than READY and IDLE.  For
+     *   example, suppose you round-robin on 2 Subchannels, in READY and CONNECTING states
+     *   respectively.  If the picker ignores the state and pick them equally, 50% of RPCs will be
+     *   stuck in buffered state until both Subchannels are READY.</li>
+     *   <li>This can also happen if you don't create a new picker at key state changes of
+     *   Subchannels.  Take the above round-robin example again.  Suppose you do pick only READY and
+     *   IDLE Subchannels, and initially both Subchannels are READY.  Now one becomes IDLE, then
+     *   CONNECTING and stays CONNECTING for a long time.  If you don't create a new picker in
+     *   response to the CONNECTING state to exclude that Subchannel, 50% of RPCs will hit it and
+     *   be buffered even though the other Subchannel is READY.</li>
+     * </ul>
+     * In order to prevent unnecessary delay of RPCs, the rules of thumb are:
+     * <ol>
+     *   <li>The picker should only pick Subchannels that are known as READY or IDLE.  An exception
+     *   is that if there is only one Subchannel at a time, you may also include CONNECTING, because
+     *   the failure mode described above applies only when there are more than one Subchannels.
+     *   </li>
+     *   <li>Whenever a Subchannel that was known to be in a state that your picker would pick has
```

Comment:
Wouldn't the IDLE case mentioned above still have a problem? When the IDLE subchannel becomes ready, we need a call to updatePicker to re-assign the RPCs. That does cause load imbalance when starting though. Hmm..